pkg/dns/service.go

```diff
@@ -431,6 +434,48 @@ func (s *Service) PatchTargets(ctx context.Context, targets, hosts []string, clu
 	return nil
 }
 
+func (s *Service) DNSDeletion(ctx context.Context, owner interface{}) error {
+	gateway, ok := owner.(*gatewayv1beta1.Gateway)
+	if ok {
+		trafficAccessor := traffic.NewGateway(gateway)
+		allHosts := trafficAccessor.GetHosts()
+
+		for _, host := range allHosts {
+			managedZone, subDomain, err := s.GetManagedZoneForHost(ctx, host, trafficAccessor)
+			if err != nil {
+				return err
+			}
+			if managedZone == nil {
+				log.Log.Info("Managed zone not found cannot get DNS record")
+				return nil
+			}
+
+			log.Log.Info("getting DNS record associated with deleting gateway")
+			dnsRecord, err := s.GetDNSRecord(ctx, subDomain, managedZone)
+			if err != nil && !k8serrors.IsNotFound(err) {
+				return err
+			}
+			if dnsRecord == nil {
+				log.Log.Info("DNS record not found", "Host name:", host)
+				continue
+			}
+			err = s.controlClient.Delete(ctx, dnsRecord)
+			log.Log.Info("DNS record deleted ", "DNSRecord:", dnsRecord.Name)
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}
+	// TODO ingress case
+	_, ok = owner.(*v12.Ingress)
+	if ok {
+		return nil
+	}
```

Comment:
If you just pass the traffic.Interface into this method you won't need to worry about this.

Reply:
Not sure that Gateway implements `traffic.interface` :thinking:

Reply:
You are already getting here https://github.com/Kuadrant/multicluster-gateway-controller/pull/162/files#diff-547de635669379196f2ffdf2e286b12a8736b8ba08e18cd156a238f2c3fb2578R440. I'm suggesting you pass that into the function from the controller that already knows what type of resource you are dealing with.

Reply:
implemented